modules/distributor/forwarder.go

```diff
@@ -165,20 +190,38 @@ func (f *forwarder) watchOverrides() {
 
 			// Spawn a goroutine to asynchronously shut down queue managers
 			go func() {
-				for _, qm := range queueManagersToDelete {
+				for _, q := range queuesToDelete {
 					// shutdown the queue manager
 					// this will block until all workers have finished and the queue is drained
-					level.Info(log.Logger).Log("msg", "Shutting down queue manager", "tenant", qm.tenantID)
-					if err := qm.shutdown(); err != nil {
-						level.Error(log.Logger).Log("msg", "error shutting down queue manager", "tenant", qm.tenantID, "err", err)
+					_ = level.Info(f.logger).Log("msg", "Shutting down queue manager", "tenant", q.TenantID())
+					if err := q.Shutdown(context.Background()); err != nil {
+						_ = level.Error(f.logger).Log("msg", "error shutting down queue manager", "tenant", q.TenantID(), "err", err)
 					}
 				}
 			}()
 
 			// Synchronously update queue managers
-			for _, qm := range queueManagersToAdd {
-				level.Info(log.Logger).Log("msg", "Updating queue manager", "tenant", qm.tenantID)
-				f.queueManagers[qm.tenantID] = newQueueManager(qm.tenantID, qm.queueSize, qm.workerCount, f.forwardFunc)
+			for _, q := range queuesToAdd {
+				_ = level.Info(f.logger).Log("msg", "Updating queue manager", "tenant", q.tenantID)
+
+				processFunc := func(ctx context.Context, data *request) error {
```

Comment:
this function exists twice. instead of declaring and using it locally maybe just create a normal function they both reference?

honestly this whole block (lines 207 -> 224) is the same as above. maybe consolidate the whole thing into a function to improve readability?